### PR TITLE
gui/ime: Fix respect aspect ratio.

### PR DESCRIPTION
**Pull request by Zangetsu38**
# About
- gui/ime: Fix respect aspect ratio.
Improve Grab size.

# Result
image
image

### CODE REVIEW

**Comment by Macdu** [comment]
This doesn't feel beneficial, I don't see a real reason not to make the overlay keyboard take all the screen width, except if you want the keys to be smaller. (Also the fact in the first place that why would there be a keyboard overlay when you can use the computer keyboard but anyway...)

**Comment by Zangetsu38** [comment]
> This doesn't feel beneficial, I don't see a real reason not to make the overlay keyboard take all the screen width, except if you want the keys to be smaller. (Also the fact in the first place that why would there be a keyboard overlay when you can use the computer keyboard but anyway...)

Eu what do you said ?
Is ime keyboard all is fine and offncourse i fix it to for no broken key pos and size
And ime keyboard is mandodory same size of réal render, like save data dialog
Extentnoutside broken all size and get very ugly

**Comment by Zangetsu38** [comment]
Just for explain, i juste continue fix my gui stuff for can all time keep same pos and size of element
for show that is on shell
- Shell = perfect
image

- Master = Broken
image

Keyboard to go outside ingame render with so deforme and broken initial size for all button

- Other resolution
image
image

- And so with pr is perfect
image
image

Change resolution never brokeb size of button and keep all time perfectly rendered

So of course is Benefical, is a FIX like i have done on other
i just continue fix and improve my proper interface

**Comment by Macdu** [comment]
Stop saying a change like that is a fix and is only beneficial. This is completely personal and personally I prefer the screenshot with the keyboard screen being as big as possible, I don't think everyone would want with a mouse to click some small screen areas. The Android keyboard or Windows visual keyboard always stretches its width as much as possible.

I would like to have the opinion of other people before approving this PR.

**Comment by djtuBIG-MaliceX** [comment]
I'm inclined to agree with Macdu, though if it's something of a personal preference, then it'd be better if this was **configurable** rather than a "you'll take what we give you".

**Comment by Zangetsu38** [comment]
> Stop saying a change like that is a fix and is only beneficial. This is completely personal and personally I prefer the screenshot with the keyboard screen being as big as possible, I don't think everyone would want with a mouse to click some small screen areas. The Android keyboard or Windows visual keyboard always stretches its width as much as possible.
> 
> I would like to have the opinion of other people before approving this PR.

It's a fix though, huh...
it's exactly the same fix as the other PR.
The keyboard has a fixed size I tell you, all my Ime keyboard interface is based on a pixel-perfect copy of the Ime keyboard from psvita.

And the keyboard doesn't get bigger at all, the size of the keys deforms becomes stretched in height or width you can see it and it's horrible.

And once again, the keyboard must not come out of the ingame rendering, it is a keyboard that is ingame, and which must therefore remain in the viewport area

In short, this pr is literally the same thing as the others, so a fix so that it keeps this size and position regardless of the resolution of the window.

**Comment by Zangetsu38** [comment]
> I'm inclined to agree with Macdu, though if it's something of a personal preference, then it'd be better if this was **configurable** rather than a "you'll take what we give you".

there is no personal option to know here precisely...
I fix my own interface, it stops there
just see the comparative screens to understand the current problem, it's simple though
I have cooded this alone, i perfectly know why i done this

**Comment by djtuBIG-MaliceX** [comment]
Recommend get someone with a Steam Deck to test this out ;)

**Comment by Zangetsu38** [comment]
The keyboard will pop out of the screen, which already is horrible, it's basic like the save data dialog, and must be rendered only in the viewport, the ime is an ingame module, just have to see the shell screen huh. ..

And so depending on the resolution the keys become deformed and can therefore be crushed as you see on one of the screens.

In the end, you just have to look at the screen of the shell, and of my pr to see that it is indeed a real fix, eh, and so yes it is just beneficial.

Since when is it good to have everything trained that no longer respects this basic size?
it's just ugly...

**Comment by Zangetsu38** [comment]
if for you that is better you have big probleme is your eyes 💀 
image
here is just broken
image
image
probleme is no more big, probleme is it get full deformed and totally broken
Ime should be keep all time same pos and size in all resolusion possible, if ONE resolution get brokien, want said is not fine no... 
Also, Petmac confirm me, my fix is totaly good for can keep respect size relate aspect ration like my pr so

**Comment by bookmist** [comment]
I think, that if IME keyboard would be LLEd it looks the same as in this PR. It may be not a fix, but making behavior more like real Vita, and I think it's ok to merge this PR.

**Comment by petmac** [comment]
I like the fix for a few reasons:
* It's more true to the PS Vita's IME
* The aspect ratio stays correct
* It looks OK even on wide/tall monitors

However, I'm not sure I would say it's a definitive "fix":
* Using the OS's (or SDL's) IME would be better
* It could stretch to cover the screen AND keep the correct aspect ratio, which would look good for tall monitors and not cover gameplay

**Comment by Zangetsu38** [comment]
Of course, like you 2, I would love to make the IME work in lle mode
as for the save data dialog by the way
and all the common dialogs
Unfortunately, it won't be for now...

**Comment by EXtremeExploit** [comment]
imagine the android users with chocolate bars having to tap the smaller keys just because their aspect ratio is fucked, this includes myself tbh, fucked aspect ratio, i wouldnt want to tap keys 4mm wide just because the phone is super tall but not very wide, i would like to have an option like the others say, my fingers will thank you

**Comment by Zangetsu38** [comment]
> imagine the android users with chocolate bars having to tap the smaller keys just because their aspect ratio is fucked, this includes myself tbh, fucked aspect ratio, i wouldnt want to tap keys 4mm wide just because the phone is super tall but not very wide, i would like to have an option like the others say, my fingers will thank you

I have no understund what you said...
Is totaly inversed, bouton get more smal when aspect ratio is not respected
Just look this...
image

Read what petmac explain please
Nobody ask option for this, already one exist for can stretch display area

**Comment by Macdu** [comment]
IME provides a keyboard when a game needs some text input for the user.
I don't really see a need for something looking exactly (up to the aspect ratio) like the PS Vita. From my (personal), point of view, what is needed is a method which provides the most user-friendly way to enter some input. Restricting the keyboard to a 16:9 aspect ratio is by far not user friendly and causes keys to be very small when the aspect ratio is quite different to 16:9. 
This IME keyboard is just some imgui code, it would be really easy to adapt it to all possible aspect ratio while always filling all the bottom part (it almost already is, the only issues I can see is that the letter can be sometimes slightly too big and are not always completely centered).

**Comment by Zangetsu38** [comment]
sorry, but my code here i just fix it, with get all time perferctly siz, just that all, i get so same result if using lle of ime, i have right, just is a fact

**Comment by Zangetsu38** [comment]
Just read what pet mac explain...
and for 
>This IME keyboard is just some imgui code, it would be really easy to adapt it to all possible aspect ratio while always filling all the bottom part (it almost already is, the only issues I can see is that the letter can be sometimes slightly too big and are not always completely centered).

sorry but you have wrong, is litteraly not possible here...
Again, IME code here is MY code, i have full coded it alone... stop said me how get better this, you have just wrong

All you thinks is juste not possible, if you resize window you get obligatory all size of button and keyboard totaly broken, if text get too big is normal, is just because proportion get broken, if very simple...

seriously, juste look screen of sheel, is easy to understund no ? 
only one way is respect aspect ratio, is just logical, all other is just shit and broken

**Comment by Zangetsu38** [comment]
>It could stretch to cover the screen AND keep the correct aspect ratio, which would look good for tall monitors and not cover gameplay

@petmac for this part, is not possible, like i explain, i have no idea how calculate this, but if you can succes that, for can keep size of button and of aspect ratio so on size keyboard one problem.. you obligatory hide render with so text ingame get full hiden with this more big keyboard... 
Only one fix is my pr, no any other things

**Comment by Zangetsu38** [comment]
i have done àjne testing code for show you result, and so is exactly what i have explain before
I have so succes keep same proportion of keyboard when you stretch window, for it all time take full size in width
image

And so, like i have explain, that is juste dead, is not possible
you no can using full size of widh for display keyboard, it obligatory hide render of real text inpput inside ingame
i no understund why you no understund this very simple fact.
Ime is Virtual Ingame Keyboard
it full depending of ingame size
Only this pr can FIX this issue with keyboard deformed and broken proportion, ONLY that

**Comment by Zangetsu38** [comment]
Opened this pr again after seeing an Android user complain and report the exact problem that this pr was fixing and that I had specifically made for Android and wide screens in general.

And yes once again, I repeat, it is indeed a fix for the interface, it is normal, and there is no need to leave the choice to the user, if they really want to have everything on everything. the screen, there is the option to stretch the screen which is there, which will cause exactly the same problem...
So here I am correcting just one problem, that's all.

I remind you but, I coded this interface based on the psvita keyboard, and it therefore uses the dimensions provided for a 960x544 window, or at least, this ratio
And on a PC or other 16/9 screen there is therefore no problem or very light in full screen.
image

**Review by bookmist — APPROVED** [review]
